### PR TITLE
Async implementation of LLamaExecutors

### LINKED ISSUE #829: Async implementation of LLamaExecutors
### Description

I am developing WPF application that uses LLamaSharp library and particulary LLama Executores (like `InstructExecutor` and `InteractiveExecutor`).
I expect that code 
```CSharp 
await foreach (var text in executor.InferAsync(prompt, _inferenceParams))
{
    currentResult.Content += text;
}
```
doesn't block my UI thread. However, UI freezes. 

Looks, that this happens, because InferAsync awaits `InferInternal(inferenceParams, args)` method. And InferInternal implementations in InstructExecutor and InteractiveExecutor classes are synchronous. 

As an experiment I changed the `var (result, _) = Context.NativeHandle.Decode(_embeds, LLamaSeqId.Zero, batch, ref _pastTokensCount);` line in InstructExecutor to `var (result, _) = await Task.Run(() => Context.NativeHandle.Decode(_embeds, LLamaSeqId.Zero, batch, ref _pastTokensCount));` and this solved the issue.

Do you have any plans to add Async implementations for all methods that are awaited by StatefulExecutorBase in all inhereted executors?

### PR DESCRIPTION
**Pull request by asmirnov82**
Fixes #829

### CODE REVIEW

File: LLama/LLamaContext.cs [hunk 1]
```diff
@@ -564,6 +564,35 @@ public Task<DecodeResult> DecodeAsync(LLamaBatchEmbeddings batch, CancellationTo
...
         }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="tokens"></param>
+        /// <param name="id"></param>
+        /// <param name="batch"></param>
+        /// <param name="n_past"></param>
+        /// <returns></returns>
+        public (DecodeResult, int) Decode(List<LLamaToken> tokens, LLamaSeqId id, LLamaBatch batch, ref int n_past)
```
Comment (martindevans) [inline LLama/LLamaContext.cs:575]: Is this method needed anywhere by your other changes? I don't see anywhere it's called.

**Review by martindevans — CHANGES_REQUESTED** [review]
Looks good overall, just a couple of minor nits :)

**Comment by martindevans** [comment]
You can ignore the spell check failure, it's from a different change that got merged with incorrect spelling.

File: LLama/LLamaContext.cs (hunk 1 above)
Comment (asmirnov82) [reply LLama/LLamaContext.cs:575]: No, it's added only for the API to be symmetric and can be removed.

Comment (martindevans) [reply LLama/LLamaContext.cs:575]: Normally I'd say to keep it for the sake of API completeness. However, in this case the `n_past` style of calling `Decode` is legacy, and will be going away once we manage to replace the Executors.

Comment (asmirnov82) [reply LLama/LLamaContext.cs]: Fixed

File: LLama/LLamaContext.cs [hunk 2]
```diff
@@ -564,6 +564,35 @@ public Task<DecodeResult> DecodeAsync(LLamaBatchEmbeddings batch, CancellationTo
...
+            return NativeHandle.Decode(tokens, id, batch, ref n_past);
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="tokens"></param>
+        /// <param name="id"></param>
+        /// <param name="batch"></param>
+        /// <param name="n_past"></param>
+        /// <returns></returns>
```
[the remark this answers is not in the document]
Comment (asmirnov82) [reply LLama/LLamaContext.cs]: Fixed

File: LLama/LLamaContext.cs (hunk 1 above)
Comment (martindevans) [reply LLama/LLamaContext.cs]: Thanks

**Review by martindevans — APPROVED** [review]

**Comment by martindevans** [comment]
Thanks for this work @asmirnov82!

**Pull request merged**